source/_posts/2015-10-28-using-unix-tools.markdown

```diff
@@ -0,0 +1,414 @@
+a list like the following:
+
+```
+  post %r{/post\d+}     app/routes/dummy:25
+   get "/jobs"          app/routes/jobs.rb:2
+   put "/schedules/:id" app/routes/schedules.rb:55
+delete "/schedules/:id" app/routes/schedules.rb:60
+  post "/schedules/new" app/routes/schedules.rb:25
+   get "/schedules"     app/routes/schedules.rb:5
+```
+
+That is to say, we want to print the HTTP verb in the first column with
+right-alignment, followed by the route path and lastly the route file with the
+line number. We must also note that we want the list ordered by the route path
+(second column).
+
+To achieve this we first rearrange the fields of each line in the desired
+order. For this we will utilize another UNIX utility called `awk`.
+
+`awk` is a scanning and text-processing language. It is extremely useful when
+we want to treat each line as a sequence of fields separated by a space (or
+another separator).
+
+To reorder the fields in each line we simply have to:
+
+```sh
+$ grep -n -E '^\s*(get|post|put|delete) ' app/routes/*.rb \
+  | sed "s/'/\"/g;  s/  *//g;  s/ do//; s/:/ /" \
+  | awk '{ print $3, $4, $1, $2 }'
+```
+
+Each field is identified by its position starting from 1, hence the sequence we
+are left with is 3,4,1 and lastly 2. The combination of these commands results
+in the following:
+
+```
+get "/schedules" app/routes/schedules.rb 5
+post "/schedules/new" app/routes/schedules.rb 25
+put "/schedules/:id" app/routes/schedules.rb 55
+delete "/schedules/:id" app/routes/schedules.rb 60
+get "/jobs" app/routes/jobs.rb 2
+post %r{/post\d+} app/routes/dummy 25
+```
+
+It’s almost what we want, but first a few touch ups. To begin with, we have to
+right-align the HTTP verbs. If we take a closer look, the verb with the most
+characters is `delete` with 6. `awk` has another command similar to `print`
+called `printf`, which allows for the addition of some formatting information.
+
+```sh
+$ grep -n -E '^\s*(get|post|put|delete) ' app/routes/*.rb \
+  | sed "s/'/\"/g;  s/  *//g;  s/ do//; s/://" \
+  | awk '{ printf("%6s %s %s %s\n", $3, $4, $1, $2) }'
+
+   get "/schedules" app/routes/schedules.rb 5
+  post "/schedules/new" app/routes/schedules.rb 25
+   put "/schedules/:id" app/routes/schedules.rb 55
+delete "/schedules/:id" app/routes/schedules.rb 60
+   get "/jobs" app/routes/jobs.rb 2
+  post %r{/post\d+} app/routes/dummy 25
+```
+
+Next, we want to sort the list by the route path. To do this we will use
+another command called `sort`. This utility sorts the lines in alphabetical
+order by default, but it has a parameter that allows us to indicate which field
+from each line should be used for the comparison. Thus, indicating that we want
+to sort by the second field, the command looks like:
+
+```sh
+$ grep -n -E '^\s*(get|post|put|delete) ' app/routes/*.rb \
+  | sed "s/'/\"/g;  s/  *//g;  s/ do//; s/://" \
+  | awk '{ printf("%6s %s %s %s\n", $3, $4, $1, $2) }' \
+  | sort --key=2
```

Comment:
It makes sense that it's not ideal to sort the list. 

But the idea is to show other commands so people can see how you can mix many different commands to achieve different results. In this case it's more important to show that you can mix a sort command and sort by any column than be right about the output :smile: